File: apps/server/src/domain/sessionStore.ts
import {
  CONTEXT_WINDOW_TOKENS,
  INPUT_BUDGET_RATIO,
  RESERVED_OUTPUT_TOKENS,
  SUMMARY_TRIGGER_MESSAGE_COUNT,
} from '../memory/constants.js';
import { estimateTextTokens } from '../memory/tokenEstimate.js';
import {
  getFileSessionRepositoryOrNull,
  getSessionRepository,
  initSessionRepository,
  resolveSessionStoreMode,
} from './session/getSessionRepository.js';
import { duplicateRedisClient } from '../infra/redisClient.js';
import {
  listRecentDialoguesFromFiles,
} from './session/listRecentDialogues.js';
import {
  listRecentDialoguesFromRedis,
  mergeRecentDialogueItems,
} from './session/listRecentDialoguesFromRedis.js';
import { createEmptySession } from './session/normalizeSession.js';
import {
  recordSessionLoad,
  recordSessionSaveConflict,
  recordSessionSaveSuccess,
} from './session/sessionMetrics.js';
import { SessionConflictError } from './session/sessionConflictError.js';
import { enqueueSessionSave, flushSessionSaveQueue } from './session/sessionWriteQueue.js';
import type {
  AttachmentMeta,
  Message,
  RecentDialogueItem,
  Session,
  SessionAppendMessage,
} from './session/types.js';

export type {
  AttachmentMeta,
  Message,
  RecentDialogueItem,
  Session,
  SessionAppendMessage,
};

export { SessionConflictError } from './session/sessionConflictError.js';
export { SessionStreamLockBusyError } from './session/sessionStreamLock.js';
export { resolveSessionStoreMode } from './session/getSessionRepository.js';
export { getSessionMetrics } from './session/sessionMetrics.js';

export const initSessionStore = async (): Promise<void> => {
  await initSessionRepository();
};

export const loadSession = async (sessionId: string): Promise<Session> => {
  recordSessionLoad();
  return getSessionRepository().load(sessionId);
};

export const saveSession = async (sessionId: string, session: Session): Promise<Session> => {
  try {
    const saved = await getSessionRepository().save(sessionId, session, {
      expectedVersion: session.version,
    });
    session.version = saved.version;
    session.messages = saved.messages;
    recordSessionSaveSuccess();
    return saved;
  } catch (error) {
    if (error instanceof SessionConflictError) {
      recordSessionSaveConflict({
        sessionId: error.sessionId,
        expectedVersion: error.expectedVersion,
        actualVersion: error.actualVersion,
      });
    }

    throw error;
  }
};

/** 同 session 串行异步落盘，不阻塞 SSE 主流程。 */
export const saveSessionAsync = (sessionId: string, session: Session): void => {
  enqueueSessionSave(sessionId, session, saveSession);
};

export const waitForSessionPersist = async (sessionId: string): Promise<void> => {
  await flushSessionSaveQueue(sessionId);
};

/** @deprecated 仅 `SESSION_STORE=file` 时同步读内存；Redis 模式请用 `loadSession` */
export const getSession = (sessionId: string): Session => {
  if (resolveSessionStoreMode() === 'redis') {
    throw new Error('getSession() is unavailable when SESSION_STORE=redis; use loadSession()');
  }

  const fileRepo = getFileSessionRepositoryOrNull();
  if (!fileRepo) {
    return createEmptySession();
  }

  return fileRepo.getSessionSync(sessionId);
};

export const persistSession = (sessionId: string, session: Session): void => {
  saveSessionAsync(sessionId, session);
};

export const listMessages = async (sessionId: string): Promise<Message[]> =>
  (await loadSession(sessionId)).messages;

export const listRecentDialogues = async (limit = 10): Promise<RecentDialogueItem[]> => {
  try {
    const capped = Math.min(Math.max(Math.floor(limit), 1), 50);
    const fileItems = await listRecentDialoguesFromFiles(capped * 2);

    if (resolveSessionStoreMode() !== 'redis') {
      return fileItems.slice(0, capped);
    }

    const redis = duplicateRedisClient();
    try {
      const redisItems = await listRecentDialoguesFromRedis(redis, capped * 2);
      return mergeRecentDialogueItems(fileItems, redisItems, capped);
    } finally {
      redis.disconnect();
    }
  } catch (err) {
    console.warn('[sessionStore] listRecentDialogues 失败:', err);
    return [];
  }
};

export const getSessionSnapshot = async (sessionId: string) => {
  const session = await loadSession(sessionId);
  const messageCount = session.messages.length;
  const summaryTokensEstimate = estimateTextTokens(session.memorySummary);
  const conversationTokensEstimate = session.messages.reduce(
    (sum, message) => sum + estimateTextTokens(message.content) + 4,
    0,
  );
  const budgetTokensEstimate = Math.floor(CONTEXT_WINDOW_TOKENS * INPUT_BUDGET_RATIO) - RESERVED_OUTPUT_TOKENS;
  const isSummaryThresholdReached = messageCount >= SUMMARY_TRIGGER_MESSAGE_COUNT;

  return {
    messages: session.messages,
    memorySummary: session.memorySummary,
    memorySummaryUpdatedAt: session.memorySummaryUpdatedAt,
    lastId: session.lastId,
    memoryMetrics: {
      messageCount,
      conversationTokensEstimate,
      summaryTokensEstimate,
      budgetTokensEstimate,
      summaryTriggerMessageCount: SUMMARY_TRIGGER_MESSAGE_COUNT,
      isSummaryThresholdReached,
    },
  };
};

export const appendSessionMessages = async (params: {
  sessionId: string;
  messages: SessionAppendMessage[];
}): Promise<void> => {
  const session = await loadSession(params.sessionId);
  const now = Date.now();

  const nextMessages = params.messages
    .map((message, index) => ({
      role: message.role,
      content: message.content.trim(),
      timestamp: now + index,
      attachments: message.attachments,
    }))
    .filter((message) => message.content.length > 0 || (message.attachments && message.attachments.length > 0));

  if (nextMessages.length === 0) {
    return;
  }

  session.messages.push(...nextMessages);
  session.lastId += nextMessages.length;

  await saveSession(params.sessionId, session);
};

File: apps/server_py/app/domain/session_store.py
from __future__ import annotations

import time
from typing import List

from app.domain.session.errors import SessionConflictError, SessionStreamLockBusyError
from app.domain.session.get_repository import (
    get_file_session_repository_or_null,
    get_session_repository,
    resolve_session_store_mode,
)
from app.domain.session.list_recent import list_recent_dialogues
from app.domain.session.normalize import create_empty_session
from app.domain.session.types import (
    AttachmentMeta,
    Message,
    RecentDialogueItem,
    Session,
    SessionAppendMessage,
)
from app.domain.session.write_queue import enqueue_session_save, flush_session_save_queue
from app.memory.constants import (
    CONTEXT_WINDOW_TOKENS,
    INPUT_BUDGET_RATIO,
    RESERVED_OUTPUT_TOKENS,
    SUMMARY_TRIGGER_MESSAGE_COUNT,
)
from app.memory.token_estimate import estimate_text_tokens
from pydantic import BaseModel

__all__ = [
    "AttachmentMeta",
    "Message",
    "RecentDialogueItem",
    "Session",
    "SessionAppendMessage",
    "SessionConflictError",
    "SessionStreamLockBusyError",
    "SessionSnapshot",
    "MemoryMetrics",
    "init_session_store",
    "load_session",
    "save_session",
    "save_session_async",
    "wait_for_session_persist",
    "get_session",
    "persist_session",
    "list_recent_dialogues",
    "get_session_snapshot",
    "append_session_messages",
    "resolve_session_store_mode",
]


class MemoryMetrics(BaseModel):
    messageCount: int
    conversationTokensEstimate: int
    summaryTokensEstimate: int
    budgetTokensEstimate: int
    summaryTriggerMessageCount: int
    isSummaryThresholdReached: bool


class SessionSnapshot(BaseModel):
    messages: List[Message]
    memorySummary: str
    memorySummaryUpdatedAt: int | None
    lastId: int
    memoryMetrics: MemoryMetrics


async def init_session_store() -> None:
    await get_session_repository().init()


async def load_session(session_id: str) -> Session:
    return await get_session_repository().load(session_id)


async def save_session(session_id: str, session: Session) -> Session:
    saved = await get_session_repository().save(
        session_id,
        session,
        expected_version=session.version,
    )
    session.version = saved.version
    session.messages = list(saved.messages)
    return saved


def save_session_async(session_id: str, session: Session) -> None:
    enqueue_session_save(session_id, session, save_session)


async def wait_for_session_persist(session_id: str) -> None:
    await flush_session_save_queue(session_id)


def get_session(session_id: str) -> Session:
    if resolve_session_store_mode() == "redis":
        raise RuntimeError("get_session() is unavailable when SESSION_STORE=redis; use load_session()")

    file_repo = get_file_session_repository_or_null()
    if file_repo is None:
        return create_empty_session()
    return file_repo.get_session_sync(session_id)


def persist_session(session_id: str, session: Session) -> None:
    save_session_async(session_id, session)


async def get_session_snapshot(session_id: str) -> SessionSnapshot:
    session = await load_session(session_id)
    message_count = len(session.messages)
    summary_tokens_estimate = estimate_text_tokens(session.memorySummary)
    conversation_tokens_estimate = sum(
        estimate_text_tokens(message.content) + 4 for message in session.messages
    )
    budget_tokens_estimate = int(CONTEXT_WINDOW_TOKENS * INPUT_BUDGET_RATIO) - RESERVED_OUTPUT_TOKENS
    is_summary_threshold_reached = message_count >= SUMMARY_TRIGGER_MESSAGE_COUNT

    return SessionSnapshot(
        messages=session.messages,
        memorySummary=session.memorySummary,
        memorySummaryUpdatedAt=session.memorySummaryUpdatedAt,
        lastId=session.lastId,
        memoryMetrics=MemoryMetrics(
            messageCount=message_count,
            conversationTokensEstimate=conversation_tokens_estimate,
            summaryTokensEstimate=summary_tokens_estimate,
            budgetTokensEstimate=budget_tokens_estimate,
            summaryTriggerMessageCount=SUMMARY_TRIGGER_MESSAGE_COUNT,
            isSummaryThresholdReached=is_summary_threshold_reached,
        ),
    )


# Safety cap: max messages allowed in a single append call.
MAX_APPEND_MESSAGES = 500


async def append_session_messages(session_id: str, messages: List[SessionAppendMessage]) -> None:
    session = await load_session(session_id)
    now = int(time.time() * 1000)
    next_messages: List[Message] = []

    # Truncate oversized bulk appends to prevent DoS
    effective = messages[:MAX_APPEND_MESSAGES]
    for index, message in enumerate(effective):
        content = message.content.strip()
        attachments = message.attachments
        if not content and not attachments:
            continue
        next_messages.append(
            Message(
                role=message.role,
                content=content,
                timestamp=now + index,
                attachments=attachments,
            )
        )

    if not next_messages:
        return

    session.messages.extend(next_messages)
    session.lastId += len(next_messages)
    await save_session(session_id, session)
